File: webForActors/src/main/java/connQak/ConnQakTcp.java
package connQak;

import unibo.actor22comm.interfaces.Interaction2021;
import unibo.actor22comm.tcp.TcpClientSupport;
import unibo.actor22comm.utils.ColorsOut;

public class ConnQakTcp extends ConnQakBase{
    private Interaction2021 conn;

    @Override
    public Interaction2021 createConnection(String hostAddr, int port){
        try {
            conn = TcpClientSupport.connect(hostAddr,port,10);
            ColorsOut.outappl("createConnection DONE:" + conn, ColorsOut.BLUE);
            createInpurReader(conn);
            return conn;
        } catch (Exception e) {
            ColorsOut.outerr("createConnection ERROR:" + e.getMessage() );
            return null;
        }
    }



    @Override
    public void forward(String msg) {
        try {
            ColorsOut.outappl("doMove:" + msg   , ColorsOut.BLUE);
            conn.forward(msg );
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @Override
    public void request(String msg) {

    }

    @Override
    public void emit(String msg) {

    }

//-------------------------------------------------
    protected void createInpurReader(Interaction2021 conn){
        new Thread(){
            public void run(){
                try {
                    ColorsOut.out( "createInpurReader | STARTS conn=" + conn, ColorsOut.BLUE );
                    while( true ) {
                        //ColorsOut.out(name + " | waits for message  ...");
                        String msg = conn.receiveMsg();
                        ColorsOut.outappl("  | createInpurReader received:" + msg, ColorsOut.YELLOW );
                        if( msg == null ) {
                            conn.close();
                            break;
                        } else{
                            //IApplMessage m = new ApplMessage(msg);
                            //handler.elaborate( m, conn ); //chiama  ctxH
                        }
                    }
                    ColorsOut.out("TcpApplMessageHandler  |  BYE", ColorsOut.BLUE   );
                }catch( Exception e) {
                    ColorsOut.outerr( "TcpApplMessageHandler | ERROR:" + e.getMessage()  );
                }

            }
        }.start();
    }
}

File: it.unibo.raspIntro2020/resources/connQak/connQakTcp.kt
package connQak

import it.unibo.supports.FactoryProtocol
import it.unibo.`is`.interfaces.protocols.IConnInteraction
import it.unibo.kactor.MsgUtil
import it.unibo.kactor.ApplMessage

class connQakTcp( hostIP : String,  port : String,  destName : String ) :
										           connQakBase(hostIP, port, destName){
	lateinit var conn   : IConnInteraction
	
	override fun createConnection( ){ //hostIP: String, port: String  
		val fp    = FactoryProtocol(null,"TCP","connQakTcp")
	    conn      = fp.createClientProtocolSupport(hostIP, port.toInt() )    
	}
	
	override fun forward( msg: ApplMessage ){
 		conn.sendALine( msg.toString()  )				
	}
	
	override fun request( msg: ApplMessage ){
 		conn.sendALine( msg.toString()  )
		//Acquire the answer	
		val answer = conn.receiveALine()
		println("connQakTcp | answer= $answer")		
	}
	
	override fun emit( msg: ApplMessage ){
 		conn.sendALine( msg.toString()  )			
	}	
}